lib/contentful_delivery/assets.ex

```diff
@@ -0,0 +1,188 @@
+
+  https://www.contentful.com/developers/docs/references/content-delivery-api/#/reference/assets
+  """
+
+  alias Contentful.{Asset, Collection, CollectionStream, Delivery, Space}
+
+  @behaviour Collection
+  @behaviour CollectionStream
+  @doc """
+  fetches one asset from a given space
+
+  ## Examples
+      space = "my_space_id"
+      {:ok, %Asset{ meta_data: %{ id: "<asset_id>"}} = asset}
+        =  Assets.fetch_one("<asset_id>", space)
+
+      # using the configured space from config/config.exs
+      {:ok, %Asset{ meta_data: %{ id: "<asset_id>"}} = asset}
+        =  Assets.fetch_one("<asset_id>")
+
+  """
+  @impl Collection
+  @spec fetch_one(
+          String.t(),
+          Space.t() | String.t(),
+          String.t() | nil,
+          String.t() | nil
+        ) ::
+          {:ok, Asset.t()}
+          | {:error, atom(), original_message: String.t()}
+          | {:error, :rate_limit_exceeded, wait_for: integer()}
+          | {:error, :unknown}
+
+  def fetch_one(asset, space \\ nil, env \\ nil, api_key \\ nil)
+
+  def fetch_one(asset_id, %Space{meta_data: %{id: space_id}}, env, api_key) do
+    space_id
+    |> build_single_request(asset_id, env, api_key)
+    |> Delivery.send_request()
+    |> Delivery.parse_response(&build_asset/1)
+  end
+
+  def fetch_one(asset_id, space_id, env, api_key) do
+    fetch_one(asset_id, %Space{meta_data: %{id: space_id}}, env, api_key)
+  end
+
+  @doc """
+  Fetches all assets for a given `Contentful.Space`.
+
+  Will take basic collection filters into account, specifically `:limit` and `:skip` to traverse and
+  limit the collection of assets.
+
+  ## Examples
+      space = "my_space_id"
+      {:ok, [%Asset{} | _]} = Assets.fetch_all([], space)
+
+      {:ok, [
+        %Asset{ meta_data: %{ id: "foobar_0"}},
+        %Asset{ meta_data: %{ id: "foobar_1"}},
+      ], total: 3} = Assets.fetch_all([limit: 2], space)
+
+
+      # using the configured space
+      {:ok, [
+        %Asset{ meta_data: %{ id: "foobar_1"}},
+        %Asset{ meta_data: %{ id: "foobar_2"}}
+      ], total: 3} = Assets.fetch_all(skip: 1)
+
+      {:ok, [
+        %Asset{ meta_data: %{ id: "foobar_0"}}
+      ], total: 3} = Assets.fetch_all(limit: 1)
+
+      {:ok, [
+        %Asset{ meta_data: %{ id: "foobar_2"}}
+      ], total: 3} = Assets.fetch_all(limit: 1, skip: 2)
+  """
+  @impl Collection
+  @spec fetch_all(
+          list(keyword()),
+          Space.t() | String.t(),
+          String.t() | nil,
+          String.t() | nil
+        ) ::
+          {:ok, list(Asset.t())}
+          | {:error, atom(), original_message: String.t()}
+          | {:error, :rate_limit_exceeded, wait_for: integer()}
+          | {:error, :unknown}
+
+  def fetch_all(
+        options \\ [],
+        space \\ Delivery.config(:space),
+        env \\ Delivery.config(:environment),
+        api_key \\ Delivery.config(:access_token)
+      )
+
+  def fetch_all(options, %Space{meta_data: %{id: id}}, env, api_key) do
```

Comment:
I believe we can uncomplicate things if we use the plain space_id from the config and not having a space struct passed around, which needs to be deconstructed every time to get to the value which is already a first class value via config.
This suggestion still works well in tandem with the proposal above (L94-96). Any objections?

Reply:
This will actually work both ways, since `fetch_all` is overloaded a couple of times:

```elixir
alias Contentful.Delivery.Assets

# will use the space, env and api_key configured in config/config.exs
Assets.fetch_all() 

# will fetch the space given, with an env and a custom api_key
Assets.fetch_all([], "my_space_id", "staging", "another_api_token")
```